### PR TITLE
Enable differentiating through actors with Tensorflow eager.

### PR DESCRIPTION
**Pull request by visatish**
Working version of model parallelism in TensorFlow with Ray Actors.

TODOS:
1) Parallelize backwards pass for gradient computation.
2) Fix issue of `0.0` returned instead of `None` for gradients w.r.t unused sources.

### CODE REVIEW

File: python/ray/tests/test_tf_differentiable.py [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+        def cube(self, x):
+            return x**3
+
+        def double(self, x):
+            return 2 * x
+
+
+        # multiple inputs, single output
+        def sum(self, *inputs):
+            return reduce(lambda x, y: x + y, inputs)
```
Comment (richardliaw) [inline python/ray/tests/test_tf_differentiable.py]: why not just `sum(inputs)`?

Comment (richardliaw) [reply python/ray/tests/test_tf_differentiable.py]: generally I think reduce is avoided :) https://stackoverflow.com/questions/13638898/how-to-use-filter-map-and-reduce-in-python-3

File: python/ray/tests/test_tf_differentiable.py [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+
+import pytest
+import tensorflow as tf
+import numpy as np
+
+import ray
+from ray.experimental.tf_utils import tf_differentiable
+
+@pytest.fixture
+def start_ray():
```
Comment (robertnishihara) [inline python/ray/tests/test_tf_differentiable.py]: use one of the fixtures from `python/ray/tests/conftest.py`.

File: python/ray/tests/test_tf_differentiable.py [hunk 3]
```diff
@@ -0,0 +1,623 @@
+from functools import reduce
+
+import pytest
+import tensorflow as tf
+import numpy as np
```
Comment (robertnishihara) [inline python/ray/tests/test_tf_differentiable.py]: these imports (the non-ray ones) should be in alphabetical order

File: python/ray/tests/test_tf_differentiable.py [hunk 4]
```diff
@@ -0,0 +1,623 @@
+from functools import reduce
```
Comment (robertnishihara) [inline python/ray/tests/test_tf_differentiable.py]: every Python file should start with

```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

File: examples/tf_model_parallelism/example.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+import numpy as np
```
Comment (robertnishihara) [inline examples/tf_model_parallelism/example.py]: Let's get rid of this file. We can include it in `examples/` when we're ready to advertise the feature.

File: python/ray/experimental/tf_utils.py [hunk 6]
```diff
@@ -195,3 +200,353 @@ def set_weights(self, new_weights):
...
+
+
+DEBUG = 0
+def logger_debug_info(txt):
+    if DEBUG:
+        logger.info(txt)
+
+banner_half = "".join(["#"] * 10)
+
+def tf_differentiable(num_return_vals):
```
Comment (robertnishihara) [inline python/ray/experimental/tf_utils.py]: The important functions in this file need a lot of documentation to explain precisely what they're doing (so that someone else could deeply understand it and modify the code).

File: python/ray/experimental/tf_utils.py [hunk 7]
```diff
@@ -195,3 +200,353 @@ def set_weights(self, new_weights):
                 for (name, value) in new_weights.items()
                 if name in self.placeholders
             })
+
+
+DEBUG = 0
+def logger_debug_info(txt):
```
Comment (robertnishihara) [inline python/ray/experimental/tf_utils.py]: Let's get rid of code like this in preparation for merging it.

File: python/ray/experimental/tf_utils.py [hunk 8]
```diff
@@ -195,3 +200,353 @@ def set_weights(self, new_weights):
...
+        logger.info(txt)
+
+banner_half = "".join(["#"] * 10)
+
+def tf_differentiable(num_return_vals):
+    def tf_differentiable_helper(method):
+        #TODO(vsatish): Can we automatically determine 
+        #that a method is differentiable at runtime based on inputs?
+
+#        def differentiable_method(self, identifier, forward_pass,
```
Comment (robertnishihara) [inline python/ray/experimental/tf_utils.py]: remove all commented out code

File: python/ray/tests/test_tf_differentiable.py [hunk 9]
```diff
@@ -0,0 +1,623 @@
...
+    @ray.remote
+    class Actor(object):
+        """ Ray Actor with TF differentiable functions. """
+
+        def __init__(self):
+            # enable eager execution
+            tf.enable_eager_execution()
+
+        # single input, single output
+        @tf_differentiable(num_return_vals=1)
```
Comment (robertnishihara) [inline python/ray/tests/test_tf_differentiable.py]: `num_return_vals` should be specified in `@ray.method(num_return_vals=...)`, not in `tf_differentiable`.

Also, 1 is the default, it should only be specified for values != 1

File: python/ray/experimental/tf_utils.py [hunk 10]
```diff
@@ -195,3 +200,353 @@ def set_weights(self, new_weights):
...
+            
+
+            if not tf.executing_eagerly():
+                #TODO(vsatish): Can we automatically 
+                #enable this? (Tricky b/c it must be done at start of program.)
+                raise RuntimeError("Tensorflow Eager execution must "
+                                   "be enabled for experimental differentiable ray " 
+                                   "functionality.")
+
+            if forward_pass:
```
Comment (robertnishihara) [inline python/ray/experimental/tf_utils.py]: The forward pass and backward pass should be implemented in two separate helper methods.

File: python/ray/experimental/tf_utils.py [hunk 11]
```diff
@@ -195,3 +200,353 @@ def set_weights(self, new_weights):
...
+def tf_differentiable(num_return_vals):
+    def tf_differentiable_helper(method):
+        #TODO(vsatish): Can we automatically determine 
+        #that a method is differentiable at runtime based on inputs?
+
+#        def differentiable_method(self, identifier, forward_pass, 
+#                                  persistent_tape, dys,
+#                                  arg_types, kwarg_types, *args, 
+#                                  **kwargs):
+        def differentiable_method(self, identifier, forward_pass,
```
Comment (robertnishihara) [inline python/ray/experimental/tf_utils.py]: This needs an informative doc string explaining the concepts and design decisions as well as the meanings of the arguments.

For doc strings throughout, please exactly follow the style in http://www.sphinx-doc.org/en/1.5/ext/example_google.html

**Review by raulchen — COMMENTED** [review]
I don't think it's a good idea to introduce tensorflow specific code in `actor.py/worker.py`. Can we do this at the app level? I haven't looked closely, but it looks that at least we can just introduce more general mechanisms (e.g., pre/post hooks) to ray core, right?

File: python/ray/tests/test_tf_differentiable.py [hunk 12]
```diff
@@ -0,0 +1,850 @@
...
+        @tf_differentiable
+        def two_in_square_cube_v2(self, x, y):
+            return x**2, x**3
+
+        @ray.method(num_return_vals=5)
+        @tf_differentiable
+        def f(self, x, y, z):
+            return x**2, y**3, x**4, 2 * z, x * y
+
+        @ray.method(num_return_vals=2)
```
[the remark this answers is not in the document]
Comment (visatish) [reply python/ray/tests/test_tf_differentiable.py]: When would they have different number of return values?

File: python/ray/experimental/tf_utils.py [hunk 13]
```diff
@@ -196,3 +199,421 @@ def set_weights(self, new_weights):
...
+
+            # fetch the cached TF state
+            tape, tensors, results = self.__ray_tf_info__[identifier]
+
+            # process the partial downstream gradients
+            if not isinstance(results, (list, tuple)):
+                results = [results]
+            dys_processed = [] 
+            for dy, result in zip(dys, results):
+                if isinstance(dy, np.floating) and dy == 0.0:
```
[the remark this answers is not in the document]
Comment (visatish) [reply python/ray/experimental/tf_utils.py]: It can happen in the case of a literal `0.0` gradient, right? If you mean detecting whether or not a `0.0` appeared between the downstream op's `grad` and the current `grad` that's a trickier problem.

Things do break without the `if` statement, but that's because shapes don't match up and a `reshape` op fails in TF.

File: python/ray/experimental/tf_utils.py [hunk 14]
```diff
@@ -196,3 +199,421 @@ def set_weights(self, new_weights):
...
+        Note:
+            The tensors returned here are the `graph_tensor` attributes of the `TFObjectID`s returned
+            from the remote actor method invocation. They are merely to link the local TF graph.
+
+        """
+
+        # submit the remote op
+        identifier = np.random.bytes(20)
+        forward = True
+        persistent_tape = False #TODO:(vsatish) when would we want this to be True?
```
[the remark this answers is not in the document]
Comment (visatish) [reply python/ray/experimental/tf_utils.py]: Is there any way to automatically infer this without explicitly passing it in somewhere? I did a quick initial search and couldn't find anything.

File: python/ray/experimental/tf_utils.py [hunk 15]
```diff
@@ -196,3 +199,421 @@ def set_weights(self, new_weights):
...
+                results = method(self, *processed_args, **processed_kwargs)
+
+            # cache the TF state for backward pass
+            if not hasattr(self, "__ray_tf_info__"):
+                self.__ray_tf_info__ = {}
+            self.__ray_tf_info__[identifier] = (tape, tensors, results)
+
+            # process the results
+            if isinstance(results, (list, tuple)):
+                processed_results = [result.numpy() if isinstance(result, (tf.Tensor, tf.Variable))
```
[the remark this answers is not in the document]
Comment (visatish) [reply python/ray/experimental/tf_utils.py]: This is possible if we have an identity function, right?

File: python/ray/experimental/tf_utils.py [hunk 16]
```diff
@@ -218,28 +378,34 @@ def differentiable_method(self, identifier, forward_pass,
         """TensorFlow differentiable actor method wrapper.
 
         Args:
-            identifier (str): The unique byte-string identifier for this method.
+            self (ray.Actor): The ray Actor invoking this method.
+            identifier (str): The unique byte-string used to associate the 
+                              forward and backward passes of this invocation.
```
Comment (robertnishihara) [inline python/ray/experimental/tf_utils.py]: Should be indented only by four spaces

File: python/ray/experimental/tf_utils.py [hunk 17]
```diff
@@ -196,3 +198,769 @@ def set_weights(self, new_weights):
                 for (name, value) in new_weights.items()
                 if name in self.placeholders
             })
+
+
+class ValType:
```
Comment (robertnishihara) [inline python/ray/experimental/tf_utils.py]:
```suggestion
class ValType(object):
```

**Comment by visatish** [comment]
@robertnishihara latest version

**Comment by visatish** [comment]
@robertnishihara Merged in and integrated https://github.com/ray-project/ray/pull/4732.

**Comment by visatish** [comment]
@robertnishihara I think this is ready-the only build tests failing are some tune tests and the Valgrind one (not too sure about the latter).

**Comment by robertnishihara** [comment]
Jenkins, retest this please.

File: python/ray/actor.py [hunk 18]
```diff
@@ -157,7 +157,7 @@ def _remote(self, args=None, kwargs=None, num_return_vals=None):
         if num_return_vals is None:
             num_return_vals = self._num_return_vals
 
-        def invocation(args, kwargs):
+        def invocation(args, kwargs, num_return_vals=num_return_vals):
```
Comment (robertnishihara) [inline python/ray/actor.py]: Was this change necessary?

Comment (visatish) [reply python/ray/actor.py]: I needed to add this because the forward and backward invocations have a different number of return values. Is there a better way to do it?

Comment (robertnishihara) [reply python/ray/actor.py]: Yeah, that makes sense.

**Comment by richardliaw** [comment]
Feel free to reopen.
